crates/sui-core/src/validator_tx_finalizer.rs

```diff
@@ -0,0 +1,426 @@
+        agg: Weak<AuthorityAggregator<C>>,
+        exit: CancellationToken,
+    ) -> Self {
+        Self {
+            authority,
+            agg,
+            exit,
+            tx_finalization_delay: TX_FINALIZATION_DELAY,
+            finalization_timeout: FINALIZATION_TIMEOUT,
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_for_testing(
+        authority: Weak<AuthorityState>,
+        agg: Weak<AuthorityAggregator<C>>,
+        exit: CancellationToken,
+        tx_finalization_delay: Duration,
+        finalization_timeout: Duration,
+    ) -> Self {
+        Self {
+            authority,
+            agg,
+            exit,
+            tx_finalization_delay,
+            finalization_timeout,
+        }
+    }
+}
+
+impl<C> ValidatorTxFinalizer<C>
+where
+    C: Clone + AuthorityAPI + Send + Sync + 'static,
+{
+    pub async fn track_signed_transaction(&self, tx: VerifiedSignedTransaction) {
+        let authority = self.authority.clone();
+        let agg = self.agg.clone();
+        let exit = self.exit.clone();
+        let tx_finalization_delay = self.tx_finalization_delay;
+        let finalization_timeout = self.finalization_timeout;
+        tokio::spawn(async move {
+            tokio::select! {
+                _ = exit.cancelled() => {},
+                _ = tokio::time::sleep(tx_finalization_delay) => {
+                    if let (Some(authority), Some(agg)) = (authority.upgrade(), agg.upgrade()) {
+                        let tx_digest = *tx.digest();
+                        tokio::select! {
+                            res = tokio::time::timeout(finalization_timeout, Self::finalize_signed_transaction(authority, agg, tx)) => {
+                                match res {
+                                    Ok(Ok(did_run)) => {
+                                        if did_run {
+                                            debug!(?tx_digest, "Successfully finalized transaction");
+                                        }
+                                    },
+                                    Ok(Err(err)) => {
+                                        error!(?tx_digest, ?err, "Failed to finalize transaction");
+                                    }
+                                    Err(_) => {
+                                        error!(?tx_digest, "Timed out finalizing transaction");
+                                    }
+                                }
+                            },
+                            _ = exit.cancelled() => {},
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    async fn finalize_signed_transaction(
+        authority: Arc<AuthorityState>,
+        agg: Arc<AuthorityAggregator<C>>,
+        tx: VerifiedSignedTransaction,
+    ) -> anyhow::Result<bool> {
+        let epoch_store = authority.load_epoch_store_one_call_per_task();
+        if tx.auth_sig().epoch != epoch_store.epoch() {
```

Comment:
its a bit hard to be sure without seeing how we plan to call `track_signed_transaction`, but it seems that this could be simplified by using `within_alive_epoch`. You wouldn't need the exit token or the epoch check anymore.

Reply:
We will just call track_signed_transaction in authority.rs (or authority_server.rs) after we sign the transaction.